### PR TITLE
Complex points

### PR DESCRIPTION
**Pull request by behdad**
Using complex numbers to represent points.  I see a 40% speedup.  Have not tried to improve it.

### CODE REVIEW

**Comment by behdad** [comment]
Ok I think I'm done with this for now.  It's 7 times faster than it was a week ago.  That should get us going for a while.  Tests still need to be checked / fixed.

**Comment by anthrotype** [comment]
> It's 7 times faster

wow, that's good! I've never seen complex numbers used like that, but it works, so that's cool.

I guess before we merge this we should also bump the version number in setup.py, as there may be already fonts that were produced using the current algorithm.

I was wondering. Why is such a non-sampling error better the previous? Is it just faster to compute or is it also more accurate?

**Comment by jamesgk** [comment]
It's definitely more accurate. We don't decide whether all points in the curve are inside the tolerance or not until we know for sure, which is impossible with sampling. Whether it would be faster was up in the air, but in practice it seems to be significantly so.

**Comment by anthrotype** [comment]
I see. I was also wondering, have you considered also using the formula for calculating the area under a curve  to compare how well the quadratic approximation fits the cubic (i.e. the difference of the two areas is within a certain tolerance)?
I haven't tried myself. I guess that could be easier for a user who is not a programmer to understand intuitively.

**Comment by jamesgk** [comment]
Yea, we considered that. But ultimately it's no good because two curves can have the same area but still have wildly different shapes.

**Comment by anthrotype** [comment]
Actually forget what I just said. Same area does not guarantee the two curves are similar.

**Comment by behdad** [comment]
Here's some good reading on the complex-numbers approach for representing points: http://www.solitaryroad.com/c606.html

**Comment by jamesgk** [comment]
This yields the same results as my PR, so I think it's correct and we can update the tests accordingly. Not sure if you want to do that in a separate PR or this one.

A nit I have with this PR is style. Mostly variable names that are not consistent / don't conform to the style guide and a lack of horizontal whitespace (also, defining 1/27 as a keyword argument?). If you have strong feelings about it let's keep it as is. Otherwise I'm keen to go in and "clean stuff up" based on my preferences.

File: Lib/cu2qu/__init__.py [hunk 1]
```diff
@@ -16,154 +16,182 @@
...
+        p3 = p[3]
+        p1 = p0+(qp1-p0)*(2./3)
+        p2 =  p3+(qp1-p3)*(2./3)
+        if not cubic_farthest_fit(0,
+                                  p1-p[1],
+                                  p2-p[2],
+                                  0, tolerance):
             return None
-        return [p[0], p1, p[3]]
+        return p0, p1, p2, p3
```
Comment (jamesgk) [inline Lib/cu2qu/__init__.py]: This should be `return p0, qp1, p3`, right? I noticed that all the conversions on actual fonts didn't result in any length-3 splines. But it's completely bewildering to me how the tests were passing.

Comment (behdad) [inline Lib/cu2qu/__init__.py]: Oh, good catch.  I have a fox ok my laptop, will push out as soon as wifi.

That should have broken len=1 conversions, right?  Why do you so change in len-3?

Other things look good?

Comment (jamesgk) [inline Lib/cu2qu/__init__.py]: Sorry yea, I meant "length 3" as in three points, so a spline with 1 segment.

Other things look good except updating tests, which I'm about to do.

**Comment by behdad** [comment]
> A nit I have with this PR is style. Mostly variable names that are not consistent /
> don't conform to the style guide

Right.  cu2qu uses pt1, pt2, pt3, pt4 for a cubic, whereas I like p0, p1, p2, p3 much better, for a few reasons: 0. first thing zeroth, right? :D, 1. it's reflect same numbering if p is a list-like: p[0], p[1], ..., 2. in the pen protocol p0 is left out as current-point, so we start args from p1; would look weird if args start at p2.

So, If you don't have a strong preference to keep pt1,... I'll replace all with p0,...

> and a lack of horizontal whitespace

Yeah I go back and forth on that.  Generally I follow everyone else in adding a space between most tokens.  But I find long statements more readable if the multiplicative operators don't have a space.  Will sprinkle some around.

I'd also add some comments since some of inlining definitely made the code less readable.

> (also, defining 1/27 as a keyword argument?).

That one is a nasty hack I picked up, apparently from Just!  It's the only way to make the code compute 1/27. only once, without paying the extra cost of looking it up in the global scope.  Look for example in fontTools.misc.psLib, search for "int=int".  That basically binds the value to a local arg that doesn't need to be looked up everytime the function is called...

I don't mind if this one goes.  In which case do the div at function entry.
> If you have strong feelings about it let's keep it as is. Otherwise I'm keen to go in
> and "clean stuff up" based on my preferences.

Definitely no strong feelings.  Feel free to clean up, or just tell me and I'll do.

**Comment by jamesgk** [comment]
> So, If you don't have a strong preference to keep pt1,... I'll replace all with p0,...

SGTM. One other thing that stuck out was the use of variable names `p` and `P` in the same function, I'd rather call p `curve` and P `pt` or something like that.

> That one is a nasty hack I picked up, apparently from Just!  It's the only way to make the code compute 1/27. only once, without paying the extra cost of looking it up in the global scope.  Look for example in fontTools.misc.psLib, search for "int=int".  That basically binds the value to a local arg that doesn't need to be looked up everytime the function is called...
> 
> I don't mind if this one goes.  In which case do the div at function entry.

Nah, this makes sense. Maybe add a comment? Also, note that we're importing division from \__future__ so I don't think we need the decimal points when dividing.

> Definitely no strong feelings.  Feel free to clean up, or just tell me and I'll do.

I don't mind if either you or I does it. Let's just make sure we're not both working on it at once! Let me know if you've started, if not I will.

**Comment by behdad** [comment]
Since I want to add comments as well, I'll do.

**Pull request merged**